File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int islands = 0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if(grid[i][j]=='1'){
                    islands++;

                    dfs(grid, i, j);
                }
            }
        }

        return islands;
    }

    void dfs(vector<vector<char>>& grid, int startRow, int startCol){
        vector<pair<int, int>> dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        grid[startRow][startCol]='0';
        for(auto [dr, dc]: dir){
            int row = startRow + dr;
            int col = startCol + dc; 
            if(row>=0 && col >=0 && row<grid.size() && col<grid[0].size() && grid[row][col] == '1' ){
                dfs(grid, row, col);
            }
        }
    }
};

File: Data Structures & Algorithms/islands-and-treasure/submission-1.js
class Solution {
    /**
     * @param {number[][]} grid
     */
    islandsAndTreasure(grid) {
        for(let i=0; i<grid.length; i++){
            for(let j=0; j<grid[0].length; j++){
                if(grid[i][j] == 0){
                    this.dfs(grid, i, j);
                }
            }
        }
    }

    dfs(grid, startRow, startCol){
        let MAX = 2147483647;
        let dir = [[-1,0], [1, 0], [0,-1], [0,1]];
        for(let i=0;i<dir.length; i++){
            let row = startRow + dir[i][0];
            let col = startCol + dir[i][1];

            if(row>=0 && col>=0 && row<grid.length && col<grid[0].length && grid[row][col] != -1){
                let curr = grid[startRow][startCol] + 1
                if(grid[row][col] > curr){
                    grid[row][col] = curr;
                    this.dfs(grid, row, col);
                }
            }
        }
    }
}
